_posts/2018-11-20-mongodb-europe-018.md

```diff
@@ -0,0 +1,411 @@
+* Finally, I want them sorted on year in ascending order
+
+Filter all with last name  "Van Hoof" and group by lastName, firstName and year:
+<div style="text-align: center;">
+  <img src="/img/2018-11-08-mongodb-europe-2018/aggregation-pipeline-builder-compose-1.png" width="100%">
+</div>
+
+Group by year and sort in ascending order:
+<div style="text-align: center;">
+  <img src="/img/2018-11-08-mongodb-europe-2018/aggregation-pipeline-builder-compose-2.png" width="100%">
+</div>
+
+As you can see from the images above, Compass will show all the intermediary results. 
+With one push of a button you can generate a command line query or the translation to a programming language.
+Compass tells me the full aggregate will look like : 
+
+```javascript
+db.people.aggregate([{$match: {
+                      lastName : "Van Hoof"
+                    }}, {$group: {
+                      _id: { 
+                        "lastName" : "$lastName",
+                        "firstName": "$firstName",
+                        "year" : "$birthYear"
+                      },
+                      count: {
+                         $sum: 1 
+                      }
+                    }}, {$group: {
+                            "_id": {
+                                "year": "$_id.year",
+                            },
+                            "occurences": {
+                              "$push" : {
+                                  "identity": "$_id",
+                                  "count": "$count"
+                              }
+                            }
+                    }}, {$sort: {
+                      "_id.year": 1
+                    }}])
+```
+
+It would have been a lot harder to write this query manually, without the pipeline builder.
+
+# Common query mistakes
+([Tips And Tricks for Avoiding Common Query Pitfalls, Christian Kurze](https://sched.co/FmAd){:target="_blank" rel="noopener noreferrer"})
+
+### Key takeaways from this session
+
+Generally speaking, there are three major causes of query slowness:
+- Blocking operations
+- Using the `$or` operator
+- Case insensitive searches
+
+It's not uncommon that a properly tuned query delivers a factor 1000 speed-up.
+So it's definitely worth investigating.
+
+### Problem 1: blocking operations
+This happens when you use an operator that needs all the data before producing results, so results can't be streamed.
+The most common culprits are the aggregation operators such as `$sort`, `$group`, `$bucket`, `$count` and `$facet`.
+
+Possible solutions:
+- Create a compound index to support your query and make sure that the sort order in the index is the same as in your query.
+- Offload the query to a secondary member.
+- Work around the issue by using a precalculated count.
+
+### Problem 2: `$and` is fast, `$or` is slow
+Sometimes a query is fast when you use the `$and` operator but slow when you use the `$or` operator.
```

Comment:
Why? Or is this common knowledge?

Reply:
It depends on the indexes you created.
It should point out that you should create an index specifically for your query.

Reply:
> Why? Or is this common knowledge?

It's an empirical finding, a statement of fact.

The "why" is expalined in the solution: $and and $or need different supporting indices.